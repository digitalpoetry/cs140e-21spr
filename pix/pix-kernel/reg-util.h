// add whatever helpers you need here.
#include "rpi-constants.h"

static inline unsigned reg_get_sp(void) {
    unsigned sp;
    asm volatile ("mov %0, sp" : "=r" (sp));
    return sp;
}

static inline unsigned reg_get_cpsr(void) {
    unsigned cpsr;
    asm volatile ("mrs %0, cpsr" : "=r" (cpsr));
    return cpsr;
}

static inline unsigned reg_get_spsr(void) {
    unsigned spsr;
    asm volatile ("mrs %0, spsr" : "=r" (spsr));
    return spsr;
}

static inline unsigned reg_get_mode(void) {
    return reg_get_cpsr() & 0b11111;
}
