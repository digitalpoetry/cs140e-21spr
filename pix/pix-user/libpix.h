#ifndef __LIBPIX_H__
#define __LIBPIX_H__
// support.
#include <stdint.h>
#include "syscalls.h"

#define die(x) do { putk(x); sys_exit(1); } while(0)

// invoke system call <sysno>
int syscall_invoke_asm(uint32_t sysno, ...);

#define sys_putc(x)     syscall_invoke_asm(SYS_PUTC, x)
#define sys_put_hex(x)  syscall_invoke_asm(SYS_PUT_HEX, x)
void sys_exit(int code);

// user level libc.
#include <string.h>
void putk(const char *msg);
void printk(const char *fmt, ...);

#endif
