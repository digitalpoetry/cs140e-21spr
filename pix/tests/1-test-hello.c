#include "libpix.h"
#include "reg-util.h"

// fake "user program" that we use as a first step --- it's linked
// in the "kernel" address space = it's easier to interpret 
// addresses.
void notmain(void) {
    putk("hello world\n");
    printk("user: stack is roughly at %x\n", reg_get_sp());
    printk("user: cpsr=%x\n", reg_get_cpsr());
    switch(reg_get_mode()) {
    case USER_MODE: putk("USER MODE!\n"); break;
    case SUPER_MODE: putk("SUPER MODE!\n"); break;
    default: die("illegal mode\n");
    }
    sys_exit(0);
}

