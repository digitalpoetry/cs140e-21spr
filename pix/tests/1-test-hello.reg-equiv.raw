FOUND: </dev/ttyUSB0>
opened tty port </dev/ttyUSB0>.
my-install: tty-usb=</dev/ttyUSB0> program=<./1-test-hello.ptag> about to boot
TRACE:simple_boot: sending 11188 bytes, crc32=f66637ef
waiting for a start
putting code
pi: <engler: CRC checked out!>
bootloader: Done.
listening on ttyusb=</dev/ttyUSB0>
kernel: stack is roughly at: 0x7ffffe8
user_code=0x400004, prog name=<1-test-hello.bin>
TRACE:	reg hash=0xfac07451
TRACE:	spsr=0x190
TRACE:	pc = 0x400004, lr = 0x400004
TRACE:	regs[0] = 0x400004
TRACE:	regs[1] = 0x7000000
TRACE:	regs[2] = 0x0
TRACE:	regs[3] = 0x0
TRACE:	regs[4] = 0x0
TRACE:	regs[5] = 0x0
TRACE:	regs[6] = 0x0
TRACE:	regs[7] = 0x0
TRACE:	regs[8] = 0x0
TRACE:	regs[9] = 0x0
TRACE:	regs[10] = 0x0
TRACE:	regs[11] = 0x0
TRACE:	regs[12] = 0x0
TRACE:	regs[13] = 0x7000000
TRACE:	regs[14] = 0x0
TRACE:	regs[15] = 0x400004
TRACE:------------------------------------------------------
TRACE:	reg hash=0x831b8654
TRACE:	spsr=0x190
TRACE:	pc = 0x400010, lr = 0x400010
TRACE:	regs[0] = 0x400004
TRACE:	regs[1] = 0x7000000
TRACE:	regs[2] = 0x0
TRACE:	regs[3] = 0x0
TRACE:	regs[4] = 0x0
TRACE:	regs[5] = 0x0
TRACE:	regs[6] = 0x0
TRACE:	regs[7] = 0x0
TRACE:	regs[8] = 0x0
TRACE:	regs[9] = 0x0
TRACE:	regs[10] = 0x0
TRACE:	regs[11] = 0x0
TRACE:	regs[12] = 0x0
TRACE:	regs[13] = 0x7000000
TRACE:	regs[14] = 0x400008
TRACE:	regs[15] = 0x400010
TRACE:------------------------------------------------------
TRACE:	reg hash=0xfbf639bc
TRACE:	spsr=0x190
TRACE:	pc = 0x400014, lr = 0x400014
TRACE:	regs[0] = 0x400004
TRACE:	regs[1] = 0x7000000
TRACE:	regs[2] = 0x0
TRACE:	regs[3] = 0x0
TRACE:	regs[4] = 0x0
TRACE:	regs[5] = 0x0
TRACE:	regs[6] = 0x0
TRACE:	regs[7] = 0x0
TRACE:	regs[8] = 0x0
TRACE:	regs[9] = 0x0
TRACE:	regs[10] = 0x0
TRACE:	regs[11] = 0x0
TRACE:	regs[12] = 0x0
TRACE:	regs[13] = 0x6fffff8
TRACE:	regs[14] = 0x400008
TRACE:	regs[15] = 0x400014
TRACE:------------------------------------------------------
TRACE:	reg hash=0x45d5fb63
TRACE:	spsr=0x190
TRACE:	pc = 0x400018, lr = 0x400018
TRACE:	regs[0] = 0x400178
TRACE:	regs[1] = 0x7000000
TRACE:	regs[2] = 0x0
TRACE:	regs[3] = 0x0
TRACE:	regs[4] = 0x0
TRACE:	regs[5] = 0x0
TRACE:	regs[6] = 0x0
TRACE:	regs[7] = 0x0
TRACE:	regs[8] = 0x0
TRACE:	regs[9] = 0x0
TRACE:	regs[10] = 0x0
TRACE:	regs[11] = 0x0
TRACE:	regs[12] = 0x0
TRACE:	regs[13] = 0x6fffff8
TRACE:	regs[14] = 0x400008
TRACE:	regs[15] = 0x400018
TRACE:------------------------------------------------------
hello world
user: stack is roughly at 0x6fffff8
user: cpsr=0x60000190
USER MODE!
1-test-hello.bin: sys_exit(0): going to reboot
part=4
equiv values
TRACE:EQUIV:	number instructions = 903
TRACE:EQUIV:	reg hash = 0xb44f8d18
DONE!!!

Saw done

bootloader: pi exited.  cleaning up
