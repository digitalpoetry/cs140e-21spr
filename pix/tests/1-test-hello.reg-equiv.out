TRACE:	reg hash=0xfac07451
TRACE:	spsr=0x190
TRACE:	pc = 0x400004, lr = 0x400004
TRACE:	regs[0] = 0x400004
TRACE:	regs[1] = 0x7000000
TRACE:	regs[2] = 0x0
TRACE:	regs[3] = 0x0
TRACE:	regs[4] = 0x0
TRACE:	regs[5] = 0x0
TRACE:	regs[6] = 0x0
TRACE:	regs[7] = 0x0
TRACE:	regs[8] = 0x0
TRACE:	regs[9] = 0x0
TRACE:	regs[10] = 0x0
TRACE:	regs[11] = 0x0
TRACE:	regs[12] = 0x0
TRACE:	regs[13] = 0x7000000
TRACE:	regs[14] = 0x0
TRACE:	regs[15] = 0x400004
TRACE:------------------------------------------------------
TRACE:	reg hash=0x831b8654
TRACE:	spsr=0x190
TRACE:	pc = 0x400010, lr = 0x400010
TRACE:	regs[0] = 0x400004
TRACE:	regs[1] = 0x7000000
TRACE:	regs[2] = 0x0
TRACE:	regs[3] = 0x0
TRACE:	regs[4] = 0x0
TRACE:	regs[5] = 0x0
TRACE:	regs[6] = 0x0
TRACE:	regs[7] = 0x0
TRACE:	regs[8] = 0x0
TRACE:	regs[9] = 0x0
TRACE:	regs[10] = 0x0
TRACE:	regs[11] = 0x0
TRACE:	regs[12] = 0x0
TRACE:	regs[13] = 0x7000000
TRACE:	regs[14] = 0x400008
TRACE:	regs[15] = 0x400010
TRACE:------------------------------------------------------
TRACE:	reg hash=0xfbf639bc
TRACE:	spsr=0x190
TRACE:	pc = 0x400014, lr = 0x400014
TRACE:	regs[0] = 0x400004
TRACE:	regs[1] = 0x7000000
TRACE:	regs[2] = 0x0
TRACE:	regs[3] = 0x0
TRACE:	regs[4] = 0x0
TRACE:	regs[5] = 0x0
TRACE:	regs[6] = 0x0
TRACE:	regs[7] = 0x0
TRACE:	regs[8] = 0x0
TRACE:	regs[9] = 0x0
TRACE:	regs[10] = 0x0
TRACE:	regs[11] = 0x0
TRACE:	regs[12] = 0x0
TRACE:	regs[13] = 0x6fffff8
TRACE:	regs[14] = 0x400008
TRACE:	regs[15] = 0x400014
TRACE:------------------------------------------------------
TRACE:	reg hash=0x45d5fb63
TRACE:	spsr=0x190
TRACE:	pc = 0x400018, lr = 0x400018
TRACE:	regs[0] = 0x400178
TRACE:	regs[1] = 0x7000000
TRACE:	regs[2] = 0x0
TRACE:	regs[3] = 0x0
TRACE:	regs[4] = 0x0
TRACE:	regs[5] = 0x0
TRACE:	regs[6] = 0x0
TRACE:	regs[7] = 0x0
TRACE:	regs[8] = 0x0
TRACE:	regs[9] = 0x0
TRACE:	regs[10] = 0x0
TRACE:	regs[11] = 0x0
TRACE:	regs[12] = 0x0
TRACE:	regs[13] = 0x6fffff8
TRACE:	regs[14] = 0x400008
TRACE:	regs[15] = 0x400018
TRACE:------------------------------------------------------
TRACE:EQUIV:	number instructions = 903
TRACE:EQUIV:	reg hash = 0xb44f8d18
